h2o-py/h2o/model/extensions/std_coef.py

```diff
@@ -110,6 +112,8 @@ def _std_coef_plot(self, num_of_features=None, server=False):
         plt.yticks(pos[0:num_of_features], feature_labels[0:num_of_features])
         plt.tick_params(axis="x", which="minor", bottom="off", top="off",  labelbottom="off")
         plt.title("Standardized Coef. Magnitudes: H2O GLM", fontsize=20)
+        if save_plot_path is not None:
+            plt.savefig(fname=save_plot_path, **kwargs)
         # show plot
         if server:
             plt.show()
```

Comment:
@tomasfryda / @sebhrusen what do you think - should we start returning the plot object instead?

then we don't have to pass all the parameters for saving the plot - instead the caller can do it themselves

Reply:
I'm not sure if it's necessary. `plt.gcf()`[1] should return the active figure. So in theory, we could just tell users to do something like:

```python
gbm.varimp_plot(server=True)
plt.gcf().savefig("/Users/user/abc.png")
```

As long as the figure wasn't shown/closed (`server=True`), the current figure should be the last one created.

[1] https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.gcf.html

Reply:
personally, I would **always** return the plot: it offers much more flexibility/power to the end user.

@tomasfryda using `plt.gcf()` is not very natural or friendly API-wise, it's basically one of those weird property of the `plt` module which stacks various plots and then when you call a function like `varimp_plot` you have to know how many plots were created there if you want to manipulate the stack afterwards: it's breaking almost all principles of API design...
It's nice when you own all the manipulations done using `plt` but we can't rely on this when create plots on our side and expect the user to know what we did exactly.

Reply:
> using plt.gcf() is not very natural or friendly API-wise, it's basically one of those weird property of the plt module which stacks various plots and then when you call a function like varimp_plot you have to know how many plots were created there if you want to manipulate the stack afterwards: it's breaking almost all principles of API design...

I agree - it's just a hacky way that works now without any change.

Potential problem with returning figures is when we show them and then return them - some backends `close()` the figure which makes it hard to show it again (or modify and show)[1] which could make it confusing for the users. 

[1] https://stackoverflow.com/questions/31729948/matplotlib-how-to-show-a-figure-that-has-been-closed

Reply:
> Potential problem with returning figures is when we show them and then return them - some backends close() the figure

@tomasfryda any concrete example? What kind of backend does that?

I'm not sure I understand the difference between:
```python
def myfig():
   plt = …
   plt.show()
   return plt
   
fig = myfig()
```
and
```python
def myfig():
   plt = …
   plt.show()

fig = pyplot.gcf()
```
you're suggesting that depending on the chosen matplotlib backend, in the first case `fig` is sometimes already closed, whereas never in the second?
Where did the backend decide call `pyplot.close(plt)` or `pyplot.close(fig)` exactly?

Reply:
Sorry, I should have emphasized that my issue is with the `plt.show()`. In my hacky example, I use the `server=True` which doesn't `show()` the image. So to rephrase that hopefully more clearly - if we show the figure and then return it, it can cause some confusion since you might not be able to view it again.

I think I saw this behavior in some IDE like spyder/pycharm/vscode but I don't really recall the specific one, however,  the easiest way to reproduce is to use the `MacOSX` backend as it is the default one on macOS.

```python
import matplotlib
import matplotlib.pyplot as plt
matplotlib.get_backend() #=> 'MacOSX'
fig = plt.figure()
plt.scatter(range(5), range(5)) #=> <matplotlib.collections.PathCollection object at 0x12ce9c340>
plt.show() # fig.show can be used as well but it doesn't manage the GUI event-loop which makes the window partially unresponsive when used from `python` (as opposed to `IPython`)

# close the window with the figure
plt.show() # Doesn't show anything - no warning/error, no window
fig.show() # Doesn't show anything - no warning/error, no window
```

If we want to return the figure, I think we shouldn't `show()` it by default. Jupyter will plot it even if we don't `show()` it (but it doesn't close it) and for other usages like from CLI I think the user should call the `show()`.

Reply:
> I know in this case the user closed the figure so it might seem like it shouldn't be an issue for us

right, you're talking about a user action here. Whatever we do, if the user decides to close the picture window after it is shown the first time, then there will be difficulties to render it again.

> Also in past there were backends that were blocking only so the code didn't continue after it shown the figure until the user closed the window and this might also be relevant for python2 (and matplotlib2).

OK, but we never force the user to use `server=True`, it's not even the default behaviour, so I don't see the issue…

> Jupyter will plot it even if we don't show() it 

see https://github.com/h2oai/h2o-3/pull/5853#issuecomment-958028897 : we're now considering returning a container object, so the figure won't be rendered automatically by Jupyter if it's the last statement of a cell. Therefore the user keeps full control over the figure.

Reply:
> OK, but we never force the user to use server=True, it's not even the default behaviour, so I don't see the issue…

@sebhrusen Yes, in my opinion, the problem is that the default is `server=False` which makes the figure to `plt.show()`.
```python
if not server:
    plt.show()
```

So the issues would occur by default and the user would have to use `server=False`. It might be just a documentation issue (I remember spending quite a lot of time trying find out where the issue is).

> see #5853 (comment) : we're now considering returning a container object, so the figure won't be rendered automatically by Jupyter if it's the last statement of a cell. Therefore the user keeps full control over the figure.

I wouldn't be so sure. I tried the following in the Jupyter lab and it renders the figure. I am not sure if I am missing something but I have the feeling that this is not what you'd expect.
```python
import matplotlib.pyplot as plt
class Wrapper:
    def __init__(self):
        plt.figure()
        plt.scatter(range(3), range(3))

w=Wrapper()
print(42)
```
<img width="434" alt="Screen Shot 2021-11-03 at 12 29 40" src="https://user-images.githubusercontent.com/61695433/140052645-9999ea51-89d1-4b9f-94ea-8c944c86496a.png">

But as mentioned earlier, Jupyter does use by default backend that doesn't close the figure after rendering so it probably won't cause any other issues (other than potentially showing output in wrong cell as well).

Reply:
@tomasfryda I agree this parameter name (`server`) and semantic are a bit esoteric… anything but intuitive.
And unfortunately we have to live with it… except if:
```python
@deprecated_params({'server': (lambda server: {'render': not server})})
def my_plot(…, render=True):  # render: new param with better name/semantic
   if render:
      plt.show()
   return H2OPlotResult(…)
```
it should then be natural for users to switch off the `render` param if having any issue.

for Jupyter, this is interesting, I guess its backend is implicitely calling the `plt.gcf().show()` of sth like that then…
but yeah, if it's not closing anything, no issue there.